staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go

```diff
@@ -6893,16 +7021,365 @@ func TestValidateCustomResourceDefinitionValidation(t *testing.T) {
 			opts: validationOptions{
 				requireMapListKeysMapSetValidation: true,
 			},
-			wantError: false,
+		},
+		{
+			name: "valid x-kubernetes-validations for scalar element",
+			input: apiextensions.CustomResourceValidation{
+				OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]apiextensions.JSONSchemaProps{
+						"subRoot": {
+							Type: "string",
+							XValidations: apiextensions.ValidationRules{
+								{
+									Rule:    "self.startsWith('s')",
+									Message: "subRoot should start with 's'.",
+								},
+								{
+									Rule:    "self.endsWith('s')",
+									Message: "subRoot should end with 's'.",
+								},
+							},
+						},
+					},
+				},
+			},
+			opts: validationOptions{
+				requireStructuralSchema: true,
+			},
+		},
+		{
+			name: "valid x-kubernetes-validations for object",
+			input: apiextensions.CustomResourceValidation{
+				OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+					Type: "object",
+					XValidations: apiextensions.ValidationRules{
+						{
+							Rule:    "self.minReplicas <= self.maxReplicas",
+							Message: "minReplicas should be no greater than maxReplicas",
+						},
+					},
+					Properties: map[string]apiextensions.JSONSchemaProps{
+						"minReplicas": {
+							Type: "integer",
+						},
+						"maxReplicas": {
+							Type: "integer",
+						},
+					},
+				},
+			},
+			opts: validationOptions{
+				requireStructuralSchema: true,
+			},
+		},
+		{
+			name: "invalid x-kubernetes-validations with empty rule",
```

Comment:
do we really care about this one? Elsewhere we don't think. This can make tooling harder because you have add some pruning steps of empty lists and recursively propagate that to the root.

Reply:
I guess the main concern is could we accept redundant data left over in storage?

Reply:
Since other list field already do this, I'm in favor of being consistent (and any implications to storage also apply to those list fields). I also agree that it makes writing tools that output CRD schemas easier marginally easier to write; and there are lots of tools doing this.

Reply:
I'm in favor of disallowing a ValidationRule item in the list that doesn't actually specify a rule... `rule` should be non-empty